Added OpenTelemetryEnvironment

AlbertoMonteiro opened the pull request:
This method configure with default environmental variables definitions

Fixes #12

davidfowl commented:
Is there a test?

AlbertoMonteiro commented:
> Is there a test?

I will add later

nblumhardt commented:
Thanks for kicking this off, @AlbertoMonteiro!

A few initial thoughts -

 * The method you've added targets `LoggerAuditSinkConfiguration`; for this use case, it's `LoggerSinkConfiguration` that should be targeted
 * To get started, I don't think the new configuration overload should accept any parameters at all - just `WriteTo.OpenTelemetry()` should either read everything from the environment, or use the (OTel-defined) defaults for anything else
 * There are some subtleties in the spec around how the endpoint is interpreted; in our use case I think this will mean appending `/v1/logs` if the endpoint doesn't already carry that suffix; there are also `_LOGS_` versions of the environment variables that should override the generic endpoint address if present
 * Reading of the environment variables (and constructing resource attributes/headers) might be best factored out into an `OpenTelemetryEnvironmentVariables` kind of helper class, in part to keep the configuration extension method type clean, and also to facilitate testing

Super keen to see this work smoothly, let me know if I can clarify any of the above, or unblock anything that holds up the works :-)

AlbertoMonteiro commented:
Hey @nblumhardt tyvm for your review.

- I've added the `OpenTelemetryEnvironmentVariables`, let me know if you if you think that I should move it to some other folder/namespace
- Added tests to `OpenTelemetryEnvironmentVariables`
- I fixed the method, and now it targets the `LoggerSinkConfiguration` and calls the `OpenTelemetry` method sending proper arguments. I just keep the log level arguments, since I didn't find any env var from OTEL that I can get this information.

I also would like to know the best way to test the `OpenTelemetryEnvironment` method, since I didn't find unit tests for the other methods.

I also didn't understand entirely the stuff around the path ` /v1/logs`. The code I've written worked fine with .NET Aspire, I got this code from @davidfowl in another issue (https://github.com/serilog/serilog-aspnetcore/issues/359#issuecomment-2069731960)

nblumhardt commented:
Thanks for the updates @AlbertoMonteiro!

I need to give a little more thought to how this should fit in with the existing `WriteTo.OpenTelemetry()` methods before digging into all of the implementation details, just down to trivia otherwise.

For example, we could avoid the need for an additional configuration method if the existing configuration methods supported an `includeEnvironment` option:

```diff
    public static LoggerConfiguration OpenTelemetry(
        this LoggerSinkConfiguration loggerSinkConfiguration,
        string endpoint = OpenTelemetrySinkOptions.DefaultEndpoint,
        OtlpProtocol protocol = OpenTelemetrySinkOptions.DefaultProtocol,
        IDictionary<string, string>? headers = null,
        IDictionary<string, object>? resourceAttributes = null,
        IncludedData? includedData = null,
        LogEventLevel restrictedToMinimumLevel = LevelAlias.Minimum,
-        LoggingLevelSwitch? levelSwitch = null)
+        LoggingLevelSwitch? levelSwitch = null,
+        bool includeEnvironment = false)
    
    public static LoggerConfiguration OpenTelemetry(
        this LoggerAuditSinkConfiguration loggerAuditSinkConfiguration,
-        Action<OpenTelemetrySinkOptions> configure)
+        Action<OpenTelemetrySinkOptions> configure,
+        bool includeEnvironment = false)
```

(The default might end up being `true` FWIW, just sketching..)

Then, where the sink is configured, `OpenTelemetryEnvironment` would be given an opportunity to override the options that were otherwise supplied through code:

```diff
    public static LoggerConfiguration OpenTelemetry(
        this LoggerAuditSinkConfiguration loggerAuditSinkConfiguration,
-        Action<OpenTelemetrySinkOptions> configure)
+        Action<OpenTelemetrySinkOptions> configure,
+        bool includeEnvironment = false)
    {
        if (configure == null) throw new ArgumentNullException(nameof(configure));
        var options = new OpenTelemetrySinkOptions();
        
        configure(options);
+        if (includeEnvironment)
+        {
+            OpenTelemetryEnvironment.Configure(options);
+        }

        var exporter = Exporter.Create(
```

We could also then aim to line up the defaults specified for the existing configuration method with the (I assume) OpenTelemetry-specified defaults.

Any thoughts?

AlbertoMonteiro commented:
@nblumhardt thanks for the detailed answer.

I like the idea of using the parameter to use env var values.

So I just added a new commit with that change.

nblumhardt commented on src/Serilog.Sinks.OpenTelemetry/Helpers/OpenTelemetryEnvironment.cs:
```diff
@@ -0,0 +1,50 @@
+    {
+        options.Endpoint = Endpoint;
+        options.Headers = Headers;
+        options.ResourceAttributes = ResourceAttributes;
```
When `OTEL_EXPORTER_OTLP_ENDPOINT` is provided, and the protocol is determined to be http/protobuf, if the value doesn't end with `/v1/logs` it will be necessary to tack this on. (In a near-future version of the sink this will be fixed, but some more pieces need to come together around it.)

nblumhardt commented on src/Serilog.Sinks.OpenTelemetry/Helpers/OpenTelemetryEnvironment.cs:
```diff
@@ -0,0 +1,50 @@
+            if (part.Split('=') is { Length: 2 } parts)
+                configs.Add(parts[0], parts[1]);
+            else
+                throw new InvalidOperationException($"Invalid header format: {part}");
```
Nit: exception messages should end with `.`, although it's a bit awkward here.

AlbertoMonteiro replied:
I've added the header name to make debugging easier.

About the exception here, since the value can come from manual creation, I think that being able to know when the value is not in the expected format can help to fix it asap.

But if you still think that it isn't necessary here, I can remove it.

nblumhardt commented on src/Serilog.Sinks.OpenTelemetry/Helpers/OpenTelemetryEnvironment.cs:
```diff
@@ -0,0 +1,50 @@
+            if (part.Split('=') is { Length: 2 } parts)
+                configs.Add(parts[0], parts[1]);
+            else
+                throw new InvalidOperationException($"Invalid header format: {part}");
```
Perhaps including the environment variable name in the exception message would make debugging easier here?

nblumhardt commented on src/Serilog.Sinks.OpenTelemetry/Helpers/OpenTelemetryEnvironment.cs:
```diff
@@ -0,0 +1,50 @@
+            if (part.Split('=') is { Length: 2 } parts)
+                configs.Add(parts[0], parts[1]);
+            else
+                throw new InvalidOperationException($"Invalid resourceAttributes format: {part}");
```
As above

nblumhardt commented on src/Serilog.Sinks.OpenTelemetry/Helpers/OpenTelemetryEnvironment.cs:
```diff
@@ -0,0 +1,50 @@
+
+    private static Dictionary<string, string> GetHeaders(string? config)
+    {
+        Dictionary<string, string> configs = [];
```
Nit: variable name could be `headers`?

nblumhardt commented on src/Serilog.Sinks.OpenTelemetry/Helpers/OpenTelemetryEnvironment.cs:
```diff
@@ -0,0 +1,50 @@
+﻿using Serilog.Sinks.OpenTelemetry;
+
+namespace Serilog.Helpers;
```
Since this is OpenTelemetry-related, it can go in `Serilog.Sinks.OpenTelemetry.*`, perhaps under an `Configuration` namespace?

nblumhardt commented on src/Serilog.Sinks.OpenTelemetry/OpenTelemetryLoggerConfigurationExtensions.cs:
```diff
@@ -33,23 +34,30 @@ public static class OpenTelemetryLoggerConfigurationExtensions
     /// The `WriteTo` configuration object.
     /// </param>
     /// <param name="configure">The configuration callback.</param>
+    /// <param name="includeEnvironment">If true the configuration will be overridden with values from OTLP Exporter Configuration environment variables(https://opentelemetry.io/docs/languages/sdk-configuration/otlp-exporter/).</param>
```
```suggestion
    /// <param name="includeEnvironment">If true the configuration will be overridden with values from <see href="https://opentelemetry.io/docs/languages/sdk-configuration/otlp-exporter/">OTLP Exporter Configuration environment variables</a>.</param>
```

AlbertoMonteiro replied:
I've added this suggestion, but seems that Visual Studio doesn't show the link for param, in summary this link works fine, but in param it shows as regular text.

nblumhardt commented on test/Serilog.Sinks.OpenTelemetry.Tests/OpenTelemetryEnvironmentTests.cs:
```diff
@@ -0,0 +1,96 @@
+        var resourceAttributes = "name1=1,name2=2";
+        Environment.SetEnvironmentVariable("OTEL_EXPORTER_OTLP_ENDPOINT", endpoint);
+        Environment.SetEnvironmentVariable("OTEL_EXPORTER_OTLP_HEADERS", headers);
+        Environment.SetEnvironmentVariable("OTEL_RESOURCE_ATTRIBUTES", resourceAttributes);
```
Using the environment variables directly will cause issues if the tests rely on OTLP logging at some point in the future, and will cause tests to interfere with each other when running concurrently.

Perhaps `OpenTelemetryEnvironment` could have an internal `Configure(options, endpointVariable, headersVariable, resourceAttributesVariable)` method added, so that test can call it without setting the actual environment variables?

The "real" configure method could then just be a simple, untested:

```csharp
    public static void Configure(BatchedOpenTelemetrySinkOptions options)
    {
        Configure(
            options,
            Environment.GetEnvironmentVariable("OTEL_EXPORTER_OTLP_ENDPOINT"),
            Environment.SetEnvironmentVariable("OTEL_EXPORTER_OTLP_HEADERS"), ...
```

?

AlbertoMonteiro replied:
Thanks for the advise!

I've fixed it in another way, the Configure method receives the function that can get the env var value, so in the test, I can provide a mock function.

nblumhardt commented on src/Serilog.Sinks.OpenTelemetry/OpenTelemetryLoggerConfigurationExtensions.cs:
```diff
@@ -33,23 +34,30 @@ public static class OpenTelemetryLoggerConfigurationExtensions
         this LoggerSinkConfiguration loggerSinkConfiguration,
-        Action<BatchedOpenTelemetrySinkOptions> configure)
+        Action<BatchedOpenTelemetrySinkOptions> configure,
+        bool includeEnvironment = true)
```
If the variables are going to be considered by default, the parameter name and meaning should probably be inverted, e.g.: `ignoreEnvironment = false` (Serilog APIs follow an "optional booleans default to false" convention wherever possible.)

nblumhardt reviewed (commented):
Awesome! Thanks for making the change. I've run through and reviewed line-by-line, if any of the comments are unclear please just let me know.

AlbertoMonteiro commented:
Hey @nblumhardt I've pushed the fixes!

Let me know if still need some changes

nblumhardt commented:
Looks great, thank you @AlbertoMonteiro :tada:

AlbertoMonteiro commented:
Thanks for your support @nblumhardt !

theperm commented:
I think this PR is welcome but has missed the mark a little. It only caters for the default OTEL env vars and misses the signal specific ones for logs. These are

OTEL_EXPORTER_OTLP_LOGS_ENDPOINT="http://localhost:5341/ingest/otlp/v1/logs"
OTEL_EXPORTER_OTLP_LOGS_PROTOCOL=http/protobuf
OTEL_EXPORTER_OTLP_LOGS_HEADERS="xxxxxx"

If i want my metrics and traces to go to the defaults but i want the logs to go to Seq I can not do this at the LOG level.
I will have to override both metrics and traces signal env vars and have lofs use the defaults - this would be undesirable.
